pkg/balancer/vhost_server.go

```diff
@@ -0,0 +1,96 @@
+package balancer
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	vhost "github.com/inconshreveable/go-vhost"
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/util"
+)
+
+// VhostServer uses TLS virtual hosting using SNI to direct requests to
+// an appropriate backend.
+type VhostServer struct {
+	muxed       map[string]Strategy
+	monitor     Monitor
+	logger      *logging.Logger
+	toQuit      []chan struct{}
+	strategyMux sync.Mutex
+}
+
+func NewVhostServer(monitor Monitor, logger *logging.Logger) *VhostServer {
+	return &VhostServer{make(map[string]Strategy), monitor, logger, make([]chan struct{}, 0), sync.Mutex{}}
+}
+
+func (m *VhostServer) Serve(port int) error {
+	l, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		return util.Errorf("Could begin listening: %s", err)
+	}
+	if err != nil {
+		return util.Errorf("Could not init muxer on port %d: %s", port, err)
+	}
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			m.logger.WithField("err", err).Warnln("Connection failed")
+		}
+		go m.handleConnection(conn)
+	}
+}
+
+func (m *VhostServer) handleConnection(conn net.Conn) {
+	defer conn.Close()
+	tlsConn, err := vhost.TLS(conn)
+	if err != nil {
+		m.logger.WithField("err", err).Errorln("Failed to recognize a TLS connection")
+		return
+	}
+	defer tlsConn.Close()
+	service := tlsConn.Host()
+	strategy, err := m.determineStrategy(service)
+	err = strategy.Route(DuplexConnection(tlsConn))
+	if err != nil {
+		m.logger.WithField("err", err).Errorln("Failed to route connection")
+	}
+}
+
+// Find an existing strategy or initialize a new one for the given service
+func (m *VhostServer) determineStrategy(service string) (Strategy, error) {
+	m.strategyMux.Lock()
+	defer m.strategyMux.Unlock()
+	strategy, ok := m.muxed[service]
+	if !ok {
+		var err error
+		strategy, err = NewLeastConnectionsStrategy([]string{})
+		if err != nil {
+			return nil, err
+		}
+		// Wait for 15 seconds for routability while starting the monitoring process
+		monitorQuitCh := make(chan struct{})
+		routableCh := make(chan struct{})
+		routableErrCh := make(chan error)
+		go func() {
+			err := strategy.Routable(15 * time.Second)
+			if err != nil {
+				routableErrCh <- err
+			} else {
+				routableCh <- struct{}{}
+			}
+		}()
+		go m.monitor.MonitorHosts(service, strategy, monitorQuitCh)
```

Comment:
this is the only place where service will be a hostname instead of just a consul service name right? Probably better to do the strings.Split() herre instead of in the MonitorHosts method